From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] inetcomm: message options and mhtml absolute URLs (2019)

inetcomm: message options and mhtml absolute URLs (2019)

From Wine's todo_wine blocks, which record Windows: IMimeMessage keeps its
BOOL and ULONG options (SetOption / GetOption, a value of the wrong type
is accepted and ignored, ids that are not options are refused); the mhtml
protocol's CombineUrl with a relative URL that is itself an mhtml URL gives
that URL, or just its file when it has no location.
---
--- a/dlls/inetcomm/mimeole.c
+++ b/dlls/inetcomm/mimeole.c
@@ -1937,8 +1937,28 @@
 
     struct list body_tree;
     DWORD next_index;
+
+    /* the scalar options (OID_*), by option number */
+    DWORD option_values[TYPEDID_ID(OID_SECURITY_2KEY_CERT_BAG_64) + 1];
 } MimeMessage;
 
+/* the BOOL and ULONG options the message keeps */
+static BOOL message_option_known(const TYPEDID oid)
+{
+    switch (oid)
+    {
+    case OID_ALLOW_8BIT_HEADER: case OID_CBMAX_HEADER_LINE: case OID_SAVE_FORMAT: case OID_WRAP_BODY_TEXT:
+    case OID_CBMAX_BODY_LINE: case OID_TRANSMIT_BODY_ENCODING: case OID_TRANSMIT_TEXT_ENCODING:
+    case OID_GENERATE_MESSAGE_ID: case OID_HIDE_TNEF_ATTACHMENTS: case OID_CLEANUP_TREE_ON_SAVE:
+    case OID_BODY_REMOVE_NBSP: case OID_DEFAULT_BODY_CHARSET: case OID_DEFAULT_HEADER_CHARSET:
+    case OID_DBCS_ESCAPE_IS_8BIT: case OID_SECURITY_TYPE: case OID_SECURITY_HWND_OWNER:
+    case OID_HEADER_RELOAD_TYPE: case OID_CAN_INLINE_TEXT_BODIES: case OID_SHOW_MACBINARY:
+    case OID_SAVEBODY_KEEPBOUNDARY:
+        return TRUE;
+    }
+    return FALSE;
+}
+
 static inline MimeMessage *impl_from_IMimeMessage(IMimeMessage *iface)
 {
     return CONTAINING_RECORD(iface, MimeMessage, IMimeMessage_iface);
@@ -2725,24 +2745,20 @@
     if(pValue->vt != TYPEDID_TYPE(oid))
     {
         WARN("Called with vartype %04x and oid %08lx\n", pValue->vt, oid);
+#ifndef SG_MUTANT_OPTION_WRONGTYPE
         return S_OK;
+#endif
     }
 
-    switch(oid)
+    if (message_option_known(oid))
+    {
+        MimeMessage *This = impl_from_IMimeMessage(iface);
+#ifndef SG_MUTANT_MSG_OPTION_STORE
+        This->option_values[TYPEDID_ID(oid)] = TYPEDID_TYPE(oid) == VT_BOOL ? pValue->boolVal : pValue->ulVal;
+#endif
+    }
+    else
     {
-    case OID_HIDE_TNEF_ATTACHMENTS:
-        FIXME("OID_HIDE_TNEF_ATTACHMENTS (value %d): ignoring\n", pValue->boolVal);
-        break;
-    case OID_SHOW_MACBINARY:
-        FIXME("OID_SHOW_MACBINARY (value %d): ignoring\n", pValue->boolVal);
-        break;
-    case OID_SAVEBODY_KEEPBOUNDARY:
-        FIXME("OID_SAVEBODY_KEEPBOUNDARY (value %d): ignoring\n", pValue->boolVal);
-        break;
-    case OID_CLEANUP_TREE_ON_SAVE:
-        FIXME("OID_CLEANUP_TREE_ON_SAVE (value %d): ignoring\n", pValue->boolVal);
-        break;
-    default:
         FIXME("Unhandled oid %08lx\n", oid);
         hr = MIME_E_INVALID_OPTION_ID;
     }
@@ -2755,8 +2771,25 @@
     const TYPEDID oid,
     LPPROPVARIANT pValue)
 {
-    FIXME("(%p)->(%08lx, %p)\n", iface, oid, pValue);
-    return E_NOTIMPL;
+    MimeMessage *This = impl_from_IMimeMessage(iface);
+
+    TRACE("(%p)->(%08lx, %p)\n", iface, oid, pValue);
+
+    if (TYPEDID_ID(oid) < TYPEDID_ID(OID_ALLOW_8BIT_HEADER) || TYPEDID_ID(oid) > TYPEDID_ID(OID_SECURITY_2KEY_CERT_BAG_64))
+        return MIME_E_INVALID_OPTION_ID;
+#ifndef SG_MUTANT_OPTION_UNDEFINED
+    if (!message_option_known(oid))
+    {
+        FIXME("Unhandled oid %08lx\n", oid);
+        return TYPEDID_TYPE(oid) & VT_VECTOR ? E_NOTIMPL : MIME_E_INVALID_OPTION_ID;
+    }
+#endif
+
+    PropVariantInit(pValue);
+    pValue->vt = TYPEDID_TYPE(oid);
+    if (pValue->vt == VT_BOOL) pValue->boolVal = This->option_values[TYPEDID_ID(oid)];
+    else pValue->ulVal = This->option_values[TYPEDID_ID(oid)];
+    return S_OK;
 }
 
 /*** IMimeMessage methods ***/
@@ -3087,6 +3120,7 @@
     This->IMimeMessage_iface.lpVtbl = &MimeMessageVtbl;
     This->ref = 1;
     This->stream = NULL;
+    memset(This->option_values, 0, sizeof(This->option_values));
     list_init(&This->body_tree);
     This->next_index = 1;
 
--- a/dlls/inetcomm/protocol.c
+++ b/dlls/inetcomm/protocol.c
@@ -652,8 +652,23 @@
         return hres;
 
     if(!wcsnicmp(pwzRelativeUrl, mhtml_prefixW, len)) {
-        FIXME("Relative URL is mhtml protocol\n");
-        return INET_E_USE_DEFAULT_PROTOCOLHANDLER;
+        /* an mhtml URL of its own: it is the result, or just its file when it has no location */
+        const WCHAR *result = pwzRelativeUrl;
+        mhtml_url_t rel;
+
+        if(SUCCEEDED(parse_mhtml_url(pwzRelativeUrl, &rel)) && !rel.location) {
+#ifndef SG_MUTANT_MHTML_ABSOLUTE
+            result = rel.mhtml;
+#endif
+        }
+        len = lstrlenW(result);
+        if(len >= cchResult) {
+            *pcchResult = 0;
+            return E_FAIL;
+        }
+        lstrcpyW(pwzResult, result);
+        *pcchResult = len;
+        return S_OK;
     }
 
     len += url.mhtml_len;
--- a/dlls/inetcomm/tests/mimeole.c
+++ b/dlls/inetcomm/tests/mimeole.c
@@ -1036,9 +1036,9 @@
     PropVariantClear(&prop);
 
     hr = IMimeMessage_GetOption(msg, OID_HIDE_TNEF_ATTACHMENTS, &prop);
-    todo_wine ok(hr == S_OK, "ret %08lx\n", hr);
-    todo_wine ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
-    todo_wine ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
+    ok(hr == S_OK, "ret %08lx\n", hr);
+    ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
+    ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
     PropVariantClear(&prop);
 
     prop.vt = VT_LPSTR;
@@ -1049,9 +1049,9 @@
     PropVariantClear(&prop);
 
     hr = IMimeMessage_GetOption(msg, OID_HIDE_TNEF_ATTACHMENTS, &prop);
-    todo_wine ok(hr == S_OK, "ret %08lx\n", hr);
-    todo_wine ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
-    todo_wine ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
+    ok(hr == S_OK, "ret %08lx\n", hr);
+    ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
+    ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
     PropVariantClear(&prop);
 
     /* Invalid property type doesn't change the value */
@@ -1063,9 +1063,9 @@
     PropVariantClear(&prop);
 
     hr = IMimeMessage_GetOption(msg, OID_HIDE_TNEF_ATTACHMENTS, &prop);
-    todo_wine ok(hr == S_OK, "ret %08lx\n", hr);
-    todo_wine ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
-    todo_wine ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
+    ok(hr == S_OK, "ret %08lx\n", hr);
+    ok(prop.vt == VT_BOOL, "vt %08x\n", prop.vt);
+    ok(prop.boolVal == TRUE, "Hide Attachments got %d\n", prop.boolVal);
     PropVariantClear(&prop);
 
     /* Invalid OID */
@@ -1464,10 +1464,10 @@
         L"mhtml:file:///c:/dir/test.mht"
     }, {
         L"mhtml:file:///c:/dir/test.mht!x-usc:http://test.org", L"mhtml:file:///d:/file.html",
-        L"file:///d:/file.html", TRUE
+        L"file:///d:/file.html"
     }, {
         L"mhtml:file:///c:/dir/test.mht!x-usc:http://test.org", L"mhtml:file:///c:/dir2/test.mht!x-usc:http://test.org",
-        L"mhtml:file:///c:/dir2/test.mht!x-usc:http://test.org", TRUE
+        L"mhtml:file:///c:/dir2/test.mht!x-usc:http://test.org"
     }, {
         L"mhtml:file:///c:/dir/test.mht!http://test.org", L"123abc",
         L"mhtml:file:///c:/dir/test.mht!x-usc:123abc"
